### ISSUE TITLE
The combination if headless server + mlts is broken on and master

### ISSUE THREAD

**Issue by lambdai** [istio/istio#18921]
master : https://github.com/istio/istio/pull/18914

@howardjohn Since it is not huge divergence between master and 1.4 I think 1.4 is impacted

**Comment by lambdai** [comment]
FYI: release-1.3 is good https://github.com/istio/istio/pull/18916

**Comment by howardjohn** [comment]
Thanks @lambdai . I disabled a feature flag related to this on your master PR, will see if that makes it pass

**Comment by howardjohn** [comment]
It does not @hzxuzhonghu can you take a look? any idea what may have broken?

**Comment by hzxuzhonghu** [comment]
The traffic from headless to itself goes directly to the inbound listener?

**Comment by hzxuzhonghu** [comment]
It first intercepted by 15006(virtualInbound), and matched by destinationPort=80's filter and routed to "inbound|80|http|headless.reachability-1-7670.svc.cluster.local" cluster, which has tls settings.

**Comment by howardjohn** [comment]
@lambdai I have a headless server + strict mtls on 1.4 and it works fine. Will try master

Config:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: httpbin
spec:
  selector:
    matchLabels:
      app: httpbin
  template:
    metadata:
      labels:
        app: httpbin
    spec:
      containers:
      - name: httpbin
        image: docker.io/kennethreitz/httpbin
        resources:
          limits:
            memory: "128Mi"
            cpu: "500m"
        ports:
        - containerPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: httpbin
spec:
  clusterIP: None
  selector:
    app: httpbin
  ports:
  - name: http
    port: 80
    targetPort: 80
```

**Comment by lambdai** [comment]
@hzxuzhonghu `and matched by destinationPort=80's filter`. This filter chain should have the tls setting and expecting tls traffic if global mtls is on, but app is talking to sidecar envoy with raw traffic. That is one source of failure.

**Comment by lambdai** [comment]
I don't fully understand if reach-ability test fails because of above

**Comment by hzxuzhonghu** [comment]
To make it clear, the inbound cluster has tls settings, but the traffic out of the app is raw.

let's take 80 svc port as an example: 

app--->inbound listener--> matched by 80 filter --> inbound route --> inbound cluster with tls setting ---> app


We can see the traffic is not tls encrypted. But envoy expects.

**Comment by lambdai** [comment]
@hzxuzhonghu Are you suggesting 
1. our mtls+headless tests should disable the app calling itself, and only reenable it once the fix is done?
2. app1 call app2 should succeeds under mtls+headless?

**Comment by hzxuzhonghu** [comment]
@lambdai, In theory, it works.

**Comment by cscetbon** [comment]
Hey @hzxuzhonghu I'm trying to use a headless service with a statefulset. mtls is globally enabled and I'm trying to disable it for that service. However my nodes can't communicate. I'm on 1.4. Could it be caused by the issue described in this ticket ?

Here is deployment file I use :

```
apiVersion: "authentication.istio.io/v1alpha1"
kind: "Policy"
metadata:
  name: "default"
  namespace: "cassandra-e2e"
spec:
  peers:
    - mtls:
        mode: PERMISSIVE
---
apiVersion: "networking.istio.io/v1alpha3"
kind: "DestinationRule"
metadata:
  name: "tls-only-native-port"
  namespace: "cassandra-e2e"
spec:
  host: "*.cassandra-e2e.svc.cluster.local"
  trafficPolicy:
    tls:
      mode: DISABLE
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: cassandra
  namespace: cassandra-e2e
  name: cassandra
spec:
  clusterIP: None
  ports:
  - name: tcp-client
    port: 9042
  - name: tcp-intra-node
    port: 7000
  - name: tcp-tls-intra-node
    port: 7001
  - name: tcp-jmx
    port: 7199
  selector:
    app: cassandra
---
apiVersion: apps/v1
kind: StatefulSet
metadata: 
  namespace: cassandra-e2e
  name: cassandra
  labels:
    app: cassandra
spec:
  serviceName: cassandra
  replicas: 2
  selector:
    matchLabels:
      app: cassandra
  template:
    metadata:
      labels:
        app: cassandra
    spec:
      terminationGracePeriodSeconds: 1800
      containers:
      - name: cassandra
        # image: orangeopensource/cassandra-image:3.11.4-8u212-0.3.1-cqlsh
        imagePullPolicy: IfNotPresent
        image: gcr.io/google-samples/cassandra:v14
        command: 
        - /usr/bin/dumb-init
        - /bin/bash 
        - -c
        - /run.sh
        ports:
        - containerPort: 7000
          name: intra-node
        - containerPort: 7001
          name: tls-intra-node
        - containerPort: 7199
          name: jmx
        - containerPort: 9042
          name: native
        resources:
          limits:
            cpu: "1"
            memory: 768Mi
          requests:
            cpu: "1"
            memory: 768Mi
        securityContext:
          capabilities:
            add:
              - IPC_LOCK
        livenessProbe:
            exec:
              command:
              - /bin/bash
              - -c
              - nodetool status
            failureThreshold: 3
            initialDelaySeconds: 120
            periodSeconds: 10
            successThreshold: 1
            timeoutSeconds: 20
        readinessProbe:
            exec:
              command:
              - /bin/bash
              - -c
              - /ready-probe.sh
            failureThreshold: 3
            initialDelaySeconds: 60
            periodSeconds: 10
            successThreshold: 1
            timeoutSeconds: 10
        env:
          - name: MAX_HEAP_SIZE
            value: 192M
          - name: CASSANDRA_SEEDS
            value: cassandra-0.cassandra.cassandra-e2e.svc.cluster.local
          - name: CASSANDRA_CLUSTER_NAME
            value: K8Demo
          - name: CASSANDRA_DC
            value: DC1-K8Demo
          - name: CASSANDRA_RACK
            value: Rack1-K8Demo
          - name: POD_NAME
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: metadata.name
          - name: POD_IP
            valueFrom:
              fieldRef:
                fieldPath: status.podIP
```

**Comment by hzxuzhonghu** [comment]
@cscetbon You disabled mtls, so i think your issue is not the same. Please file a new issue and provide more details like the errors and the configdump.

**Comment by cscetbon** [comment]
@hzxuzhonghu when MTLS is disabled are communications still proxified and that could explain why it does not work. The Gossip Protocol needs to communicate locally with the current node. But I thought disabling MTLS would just skip the configured flow, am I right ?

**Comment by hzxuzhonghu** [comment]
Without tls, it works well. We have e2e cases cover this. The traffic is on this path: `app--->inbound listener (virtualInbound)--> matched by 80 filter --> inbound route --> inbound cluster ---> app`.

**Comment by cscetbon** [comment]
Ok then I’ll create a ticket cause it doesn’t work in my tests

**Comment by cscetbon** [comment]
@hzxuzhonghu could it be that it works with http protocol but not with “server-speaks-first” protocols ? I was able to get it to work with Linkerd by using specific options to skip those protocols https://github.com/linkerd/linkerd2/issues/3404. Does it help diagnosing what's going on ?

**Comment by hzxuzhonghu** [comment]
update: There is also a bug for headless service if the port and target port is not the same, say `headless` svc, the access by `curl http://headless.reachability-1-38206.svc.cluster.local:80` can not access the expected 8090 inbound listener. Because the outbound cluster from the src is ORIGINAL_DST type.

```
apiVersion: v1
kind: Service
metadata:
  annotations:
  labels:
    app: headless
  name: headless
  namespace: reachability-1-38206
spec:
  clusterIP: None
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 8090
  - name: tcp
    port: 9090
    protocol: TCP
    targetPort: 9090
  - name: grpc
    port: 7070
    protocol: TCP
    targetPort: 7070
  selector:
    app: headless
  sessionAffinity: None
  type: ClusterIP

```

**Comment by hzxuzhonghu** [comment]
For headless service in k8s, the port and target port should be the same, otherwise user can not access the service by domain：svcPort
```
For headless Services, a cluster IP is not allocated, kube-proxy does not handle these Services, and there is no load balancing or proxying done by the platform for them. 
```
https://kubernetes.io/docs/concepts/services-networking/service/#headless-services

**Comment by cscetbon** [comment]
> Without tls, it works well. We have e2e cases cover this. The traffic is on this path: app--->inbound listener (virtualInbound)--> matched by 80 filter --> inbound route --> inbound cluster ---> app.
@hzxuzhonghu did you try the deployment file I posted ? mtls is enabled globally and I'm trying to disable it specifically for port 7000 and 7001. I also tried adding 
```
  annotations: 
    traffic.sidecar.istio.io/excludeOutboundPorts: 7000,7001
```
to the statefulset metadata but it did not help

**Comment by hzxuzhonghu** [comment]
@cscetbon Is this related? https://istio.io/faq/applications/#cassandra

**Comment by cscetbon** [comment]
@hzxuzhonghu part of it relates to it. The issue with mtls is different. But I definitely think that it should support listening on the POD IP and not only on 127.0.0.1

**Comment by hzxuzhonghu** [comment]
I agree with this. It is really hack to only redirect to 127.0.0.1 But i am not sure how we can support it. maybe @lambdai has some solutions.

**Comment by howardjohn** [comment]
we could make an annotation configure if we connect to 127 or pod ip maybe?
